### PR TITLE
Query and delete connection context atomically

### LINKED ISSUE #2059: Use after free in netebpfext when 2 CONNECT_AUTH requests come with same parameters
### Describe the bug

Currently netebpfext uses the following as key to store the connection context:
1. transport endpoint handle
2. source port
3. destination port
4. destination IP
5. protocol
6. compartment ID

There are some scenarios where 2 connections can have all the above tuples same. In such a case, currently in AUTH_CONNECT classify, we hit an issue where we try to delete a node from the list twice.

### OS information

_No response_

### Steps taken to reproduce bug

NA

### Expected behavior

The bugcheck should not happen

### Actual outcome

Machine bugchecks

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by saxena-anurag**
Fixes #2059
## Description

This PR fixes the issue described in #2059:

Currently netebpfext uses the following as key to store the connection context:

1. transport endpoint handle
2. source port
3. destination port
4. destination IP
5. protocol
6. compartment ID

There are some scenarios where 2 connections can have all the above tuples same. In such a case, currently in AUTH_CONNECT classify, we hit an issue where we try to delete a node from the list twice.

## Testing

PR #2073 adds the concurrent test for this issue.

## Documentation

NA

### CODE REVIEW

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 1]
```diff
@@ -337,7 +337,7 @@ _net_ebpf_extension_sock_addr_program_info_on_client_detach(
     _In_ const net_ebpf_extension_program_info_client_t* detaching_client)
 {
     UNREFERENCED_PARAMETER(detaching_client);
-    _net_ebpf_sock_addr_get_program_context = NULL;
+    // _net_ebpf_sock_addr_get_program_context = NULL;
```
Comment (gtrevi) [inline netebpfext/net_ebpf_ext_sock_addr.c:340]: Should this be commented, if the intent is to keep the line as a history/reference?

**Review by gtrevi — APPROVED** [review]

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 1 above)
Comment (Alan-Jowett) [reply netebpfext/net_ebpf_ext_sock_addr.c:340]: Please remove all commented out code as per coding guide.

**Review by Alan-Jowett — CHANGES_REQUESTED** [review]
Please remove commented out code.

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 2]
```diff
@@ -692,7 +692,7 @@ _net_ebpf_ext_connection_context_initialize_key(
 }
 
 static NTSTATUS
-_net_ebpf_ext_get_connection_context(
+_net_ebpf_ext_get_and_delete_connection_context(
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_addr.c:695]: you are removing the context from the list, not deleting it, at least not in this function. So please rename.

File: netebpfext/net_ebpf_ext_sock_addr.c [hunk 3]
```diff
@@ -1183,7 +1170,7 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
     }
     verdict = connection_context->value.verdict;
 
-    _net_ebpf_ext_delete_connection_context(connection_context);
+    ExFreePool(connection_context);
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_addr.c:1173]: I am assuming ExFeePool does NUL check?

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 1 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c]: Yes this was unintended. Fixed.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 2 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c]: Renamed.

File: netebpfext/net_ebpf_ext_sock_addr.c (hunk 3 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_addr.c:1173]: `ExFreePool` is called only when `_net_ebpf_ext_get_and_remove_connection_context()` succeeds. So NULL check is not needed.

**Review by shankarseal — APPROVED** [review]

**Review by gtrevi — APPROVED** [review]

**Review by matthewige — APPROVED** [review]

**Comment by dthaler** [comment]
> ## Testing
> NA

Testing is never "not applicable" for any code change.  Was this but hit by existing CI/CD low memory tests or address sanitizer tests?

**Review by Alan-Jowett — CHANGES_REQUESTED** [review]
Please provide link to PR with test coverage for this change or add test coverage to this change.

**Comment by saxena-anurag** [comment]
> > ## Testing
> > NA
> 
> Testing is never "not applicable" for any code change. Was this but hit by existing CI/CD low memory tests or address sanitizer tests?

Yes I am adding tests for the race. This is a race condition and was caught in manual testing.

**Review by shankarseal — APPROVED** [review]

**Comment by saxena-anurag** [comment]
> > > ## Testing
> > > NA
> > 
> > 
> > Testing is never "not applicable" for any code change. Was this but hit by existing CI/CD low memory tests or address sanitizer tests?
> 
> Yes I am adding tests for the race. This is a race condition and was caught in manual testing.

PR #2073 adds the concurrent test for this issue.

**Comment by saxena-anurag** [comment]
> Please provide link to PR with test coverage for this change or add test coverage to this change.

Done. Added link to PR #2073

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**